core/services/relay/evm/evm.go

```diff
@@ -518,12 +518,22 @@ func (r *Relayer) NewMedianProvider(rargs relaytypes.RelayArgs, pargs relaytypes
 	if err != nil {
 		return nil, err
 	}
-	return &medianProvider{
+
+	medianProvider := medianProvider{
 		configWatcher:       configWatcher,
 		reportCodec:         reportCodec,
 		contractTransmitter: contractTransmitter,
 		medianContract:      medianContract,
-	}, nil
+	}
+
+	chainReader, err := newChainReader(lggr, r.chain, relayConfig)
+	if err != nil {
+		medianProvider.chainReader = nil
+	} else {
+		medianProvider.chainReader = chainReader
+	}
```

Comment:
This still seems confused. With everything pushed inside of `newChainReader`, why do we need to return the error?
```suggestion
	medianProvider.chainReader = newChainReader(lggr, r.chain, relayConfig)
```
This version might have typed-nil problems without corresponding changes inside of `newChainReader`. IMHO `newChainReader` is not helpful, and inlining that code directly here (the only use) would be much clearer, and avoid all of this extra complexity.

Reply:
```suggestion
	if relayConfig.ChainReader == nil {
		lggr.Info(relaytypes.ErrorChainReaderUnsupported{})
	} else if err := validateChainReaderConfig(*relayConfig.ChainReader); err != nil {
		lggr.Error(fmt.Errorf("%w: %w", relaytypes.ErrorChainReaderInvalidConfig{}, err))
	} else {
		medianProvider.chainReader = NewChainReaderService(lggr, chain.LogPoller())
	}
```

Reply:
I wanted to keep the validation inside of the `newChainReader ` and return errors under an assumption that it will be used in multiple places in the future, also thinking that overriding the `newChainReader` when its nil is temporary so we would want to return the errors because we would want to have an option to handle those returned errors at some point.

Although I see that it also makes sense to make the code as clean as possible immediately instead of thinking ahead for a use case that doesn't exist yet.

Reply:
> This version might have typed-nil problems without corresponding changes inside of newChainReader.

Could you be more specific about what typed-nil problem you see?  There was a typed nil bug here a month or more ago that was keeping CI from passing, but I after sharing a proposed fix for it in the golang channel at CLL everyone seemed to agree it was a good solution.  If there's still a subtle problem somewhere I don't see it. The fix was just what you see here:  set it explicitly to nil, instead of having the constructor return nil and using that to initialize it.

> IMHO newChainReader is not helpful, and inlining that code directly here (the only use) would be much clearer, and avoid all of this extra complexity.

This was supposed to just be an initial skeleton, but gradually most of it ended up getting filled in.  Over the past few days I went ahead and filled in everything I'd planned to do in the next JIRA (focused on initializing ChainReader from the plugin), so you can have a more complete picture to review.  Bear in mind that the ChainReader has many more methods which will be added, we've only added the first one so far (`GetLatestValue()`).  Once we add `QueryEvents` and the `Headtracker` stuff from Mercury I think the importance of having a separate constructor will grow.

Yesterday, we did decide to merge the two constructors into one though. The only purpose of the second (exported) constructor was to make it easier for tests to create a simple `chainReader` struct without having to bother filling in all the things you'd need in production.  It was so simple though, we decided inlining it in the tests is better.  We've also added some tests focusing on the constructor which are closer to end-to-end and make sure the error handling works.

Reply:
I removed the part of my previous comment asking where the nil typed error is, as I just realized you're referring to your proposed change to go back to using nil as an error value instead of having proper errors defined.

My thoughts on the error handling:  there are too many ways already in which various errors could happen here, and it's only going to grow more complex in the future as more pieces of ChainReader get added, so I think lumping them all together would be a mistake.  I've changed them in `chainlink-commons` now to errors that can be exposed to the client side, so they can be more explicitly handled in different ways. I think the fact that we were keeping them internal to the relay was obscuring how different they are.